### PR TITLE
`@dataclass` for `DataLayer`

### PR DESCRIPTION
**Pull request by altendky**

### CODE REVIEW

**Comment by altendky** [comment]
Coverage issues are all guardrail error conditions.

File: tests/core/data_layer/test_data_rpc.py [hunk 1]
```diff
@@ -803,7 +803,11 @@ async def offer_setup_fixture(
 
 async def populate_offer_setup(offer_setup: OfferSetup, count: int) -> OfferSetup:
     if count > 0:
-        for store_setup, value_prefix in {offer_setup.maker: b"\x01", offer_setup.taker: b"\x02"}.items():
+        setups: Tuple[Tuple[StoreSetup, bytes], Tuple[StoreSetup, bytes]] = (
```
Comment (xdustinface) [inline tests/core/data_layer/test_data_rpc.py:806]: This doesn't look related here?

File: chia/data_layer/data_layer.py [hunk 2]
```diff
@@ -86,38 +93,72 @@ def server(self) -> ChiaServer:
...
-        self.server_files_location = path_from_root(self.root_path, server_files_replaced)
+
+        db_path_replaced: str = config["database_path"].replace("CHALLENGE", config["selected_network"])
+
+        self = cls(
+            config=config,
+            root_path=root_path,
+            wallet_rpc_init=wallet_rpc_init,
+            log=logging.getLogger(name if name is None else __name__),
+            _shut_down=False,
```
Comment (xdustinface) [inline chia/data_layer/data_layer.py:149]:
```suggestion
```

File: chia/data_layer/data_layer.py [hunk 3]
```diff
@@ -86,38 +93,72 @@ def server(self) -> ChiaServer:
...
+            db_path=path_from_root(root_path, db_path_replaced),
+            server_files_location=path_from_root(root_path, server_files_replaced),
+            none_bytes=bytes32([0] * 32),
+            _server=None,
```
Comment (xdustinface) [inline chia/data_layer/data_layer.py:153]:
```suggestion
```

There is no default yet but seems like you could just assign it and drop this.

File: chia/data_layer/data_layer.py [hunk 4]
```diff
@@ -86,38 +93,72 @@ def server(self) -> ChiaServer:
...
+
+        self = cls(
+            config=config,
+            root_path=root_path,
+            wallet_rpc_init=wallet_rpc_init,
+            log=logging.getLogger(name if name is None else __name__),
+            _shut_down=False,
+            db_path=path_from_root(root_path, db_path_replaced),
+            server_files_location=path_from_root(root_path, server_files_replaced),
+            none_bytes=bytes32([0] * 32),
```
Comment (xdustinface) [inline chia/data_layer/data_layer.py:152]: This _could_ be a default also, maybe even not a member at all.

File: chia/data_layer/data_layer.py [hunk 5]
```diff
@@ -62,20 +63,26 @@ async def get_plugin_info(url: str) -> Tuple[str, Dict[str, Any]]:
...
     downloaders: List[str]
     uploaders: List[str]
+    server_files_location: Path
+    initialized: bool = False
+    _data_store: Optional[DataStore] = None
+    state_changed_callback: Optional[StateChangedProtocol] = None
+    _shut_down: bool = False
+    periodically_manage_data_task: Optional[asyncio.Task[None]] = None
+    _wallet_rpc: Optional[WalletRpcClient] = None
+    _subscription_lock: Optional[asyncio.Lock] = None
```
Comment (xdustinface) [inline chia/data_layer/data_layer.py:85]:
```suggestion
    _subscription_lock: asyncio.Lock = field(default_factory=asyncio.Lock)
```

Could be ^ and you could drop the related property?

**Review by xdustinface — CHANGES_REQUESTED** [review]

File: chia/data_layer/data_layer.py (hunk 3 above)
Comment (altendky) [reply chia/data_layer/data_layer.py:151]: Could we leave this?  It looks to be pointless but also looks to be part of a common pattern that we should maybe analyse.  For example, the wallet probably shouldn't need it either.  Also the timelord etc as related to my PRs around removing the unused peer ports in the config.

File: tests/core/data_layer/test_data_rpc.py (hunk 1 above)
Comment (altendky) [reply tests/core/data_layer/test_data_rpc.py]: Hum...  I thought mypy was complaining or something.  But maybe I was just annoyed it wasn't detecting the type or...  Yep, I'll make a separate pr for this.

https://github.com/chia-network/chia-blockchain/commit/680d2833e4965968c17be673f2d7d67fc0ea676d
https://github.com/Chia-Network/chia-blockchain/pull/16257

**Review by altendky — COMMENTED** [review]
thanks

File: tests/core/data_layer/test_data_rpc.py (hunk 1 above)
Comment (altendky) [reply tests/core/data_layer/test_data_rpc.py:806]: Oh yeah, this is because `DataLayer` is no longer hashable so we need to avoid hashing it such as as a dict key.

File: chia/data_layer/data_layer.py (hunk 3 above)
Comment (xdustinface) [reply chia/data_layer/data_layer.py:151]: Not sure what the common pattern is you are referring to here or what your plans are with keeping this, but im ok if you see any sense in it. So feel free to close it but i want to mention, there are other classes where we already assign a default value via the dataclass to the optional server like proposed.

File: chia/data_layer/data_layer.py (hunk 5 above)
Comment (xdustinface) [reply chia/data_layer/data_layer.py:85]: I think its just you every time this kind of thing comes up feeling like it needs to be done in a async call for some reason because its from the async package :) But we have bunch of places where we instantiate async members (`Queue`, `Event` and even also `Lock` already) like this and there doesn't seem to be an issue with it. I don't really see a reason to add this extra optional/property thing but feel free to close this.

**Review by xdustinface — DISMISSED** [review]

File: chia/data_layer/data_layer.py (hunk 3 above)
Comment (altendky) [reply chia/data_layer/data_layer.py]: https://github.com/chia-network/chia-blockchain/commit/0cc9aa655d2eaf77e033d817aa2b20107d00dd18

I...  think I wasn't thinking straight...  sorry.

**Review by xdustinface — DISMISSED** [review]

File: chia/data_layer/data_layer.py (hunk 5 above)
Comment (altendky) [reply chia/data_layer/data_layer.py]: https://github.com/chia-network/chia-blockchain/commit/96f7bbca1238ec3e0dbc1d5dfe6532b755646626

Comment (altendky) [reply chia/data_layer/data_layer.py]: The issue is with `asyncio.Queue` being created outside the loop.

https://github.com/Chia-Network/chia-blockchain/pull/12041/files

https://github.com/Chia-Network/chia-blockchain/issues/11980

**Review by arvidn — APPROVED** [review]
